nupic/research/frameworks/dendrites/utils.py

```diff
@@ -19,78 +19,251 @@
 # http://numenta.org/licenses/
 # ----------------------------------------------------------------------
 
-import matplotlib.pyplot as plt
-import numpy as np
+import math
+
 import torch
 
 
-def plot_dendrite_activations(
-    dendritic_weights,
-    context_vectors,
-    annotation_type,
-    mask_values=None
-):
+def percent_active_dendrites(dendrite_segments, context_vectors, selection_criterion):
     """
-    Returns a heatmap of dendrite activations (given dendritic weights for a single
-    neuron and context vectors) plotted using matplotlib. Note that the user must be
-    logged in to wandb on both browser and terminal to view the resulting plots, and
-    this can be done via the following command:
-
-    $ wandb login your-login-key
-
-    :param dendritic_weights: 2D torch tensor with shape (num_dendrites, dim_context)
-    :param context_vectors: 2D torch tensor with shape (num_contexts, dim_context)
-    :param annotation_type: either "regular" or "absolute"; "regular" annotates the
-                            maxmimum activation per context, whereas "absolute"
-                            annotates the absolute maximum activation per context
-    :param mask_values: list of the routing function's mask values for the output unit
-                        corresponding to `dendritic_weights` across all contexts;
-                        unused if None
+    Returns a 3D torch tensor with shape (num_units, num_dendrites, num_categories)
+    where cell k, i, j gives the fraction of inputs in category j for which dendrite i
+    of unit k is active.
+
+    :param dendrite_segments: `DendriteSegments` object
+    :param context_vectors: iterable of 2D torch tensors with shape (num_examples,
+                            dim_context) where each 2D tensor gives a batch of context
+                            vectors from the same category
+    :param selection_criterion: the criterion for selecting which dendrites become
+                                active; either "regular" (for `GatingDendriticLayer`)
+                                or "absolute" (for `AbsoluteMaxGatingDendriticLayer`)
     """
-    assert dendritic_weights.size(1) == context_vectors.size(1)
-    assert annotation_type in ("regular", "absolute")
+    num_units, num_dendrites, _ = dendrite_segments.weights.size()
+
+    percentage_activations = torch.zeros((num_units, num_dendrites, 0))
+    for j in range(len(context_vectors)):
+        num_examples = context_vectors[j].size(0)
+
+        activations = dendrite_segments(context_vectors[j])
```

Comment:
Here, the docstring says that `context_vectors` needs to be an "iterable of 2D torch tensors with shape (num_examples, dim_context)", because we want the context vectors separated by class. That means `context_vectors[j]` will be a 2D tensor.

Reply:
I see, missed that, thanks